Abstractive Summarization using Deep Learning (#3651)
section-engineering-education/engineering-education: content/articles/abstractive-summarization-using-deep-learning/index.md

Let's install PyTorch.

Navigate to PyTorch's main [website](https://pytorch.org/). On the main webpage, you'll see an `install` option. Click on it. Here, you can choose your PyTorch build of choice, your OS, installation package, language, and compute platform. This will generate a code that you will use to install PyTorch. Please note that different selections generate different installation codes.

For our case, we'll choose the `LTS (1.8.2)` Pytorch build, `Linux` OS, the `Pip` package, `Python` programming language, and finally `CUDA 10.2`. Since we'll be using Google Colab. If you don't have a GPU in your machine, you can still go ahead and choose the CPU option.

After the selection, the following installation command is generated:

```python
!pip3 install torch==1.8.2+cu102 torchvision==0.9.2+cu102 torchaudio==0.8.2 -f https://download.pytorch.org/whl/lts/1.8/torch_lts.html
```
If you're using Jupyter notebook or Google Colab, make sure to add the `!` before the command. Otherwise, it won't install.

This command successfully installs PyTorch.

Let's install our second dependency, which is the HuggingFace transformers.

```python
!pip3 install transformers
```

> This command will successfully install the huggingface transformer library.

Reviewer: Consider removing this sentence

The library has many models, including BERT and GPT-2 models that perform various tasks, but we’ll be leveraging the pre-trained language pipeline for our purpose. Rather than going ahead and training a huge language model such as GPT-2 with 1.5 billion parameters, one can leverage the ML pipeline instead."""
```

We've picked some text from this [tutorial](/engineering-education/building-a-simple-translation-app-using-python-for-beginners/) on Section's website.

Using our tokenizer, which we loaded earlier, let's convert our text into its token representation.

```python
tokens = tokenizer_model(text, truncation=True, padding="longest", return_tensors="pt")
```
We've created a variable called `tokens` to store our token representation. We use our tokenizer `tokenizer_model`, which we created earlier, to tokenize our texts. The `truncation=True` argument allows the model to truncate our texts into a size suitable for input into our model. The `return_tensors` argument tells the model to use tensors from PyTorch.

Let's view our tokens. We achieve this by writing the following command:

```python
tokens
```

> Let's go ahead and try to summarize this text.

Reviewer: ```suggestion
It's time to summarize our text.
```

```python
summary = loaded_model.generate(**tokens)
```
The `**tokens` unpacks our tokens and pass them into our model. The asterisks in `**tokens` are simply adding the `input_ids` and `attention_mask` present in the results above. 

> Let's check our generated summary results. We do this by typing the following command:

Reviewer: ```suggestion
To check our generated summary results in tokens, we type `summary`.
```

```python
summary
```

> These results represent our output tensors. This is our summary in tokens.
> 
> These may be just a bunch of numbers to humans, but not to machines. This is how they understand language. Decoding these values will help us make sense of these numbers.

Reviewer: ```suggestion
These results represent our output tensors. These may be just a bunch of numbers to humans, but not to machines. This is how they understand language. Decoding these values will help us make sense of these numbers.
```

```python
tokenizer_model.decode(summary[0])
```
The `summary` result above shows that the results are in a nested list. But we only need the first result, so that's why we've indexed [0].

> Running that command shows our summarized text. 
> 
> Impressive, right?
> 
> All that big block of text summarized into that short text. 

Reviewer: ```suggestion
Running the above command shows our summarized text. Isn't it impressive how that block of text has been summarized?
```

> So, here's the actual test. If you go through the huge block of text, you'll find that this summarized version of the text doesn't exist. It's novel and has been completely generated by the model.

Reviewer: The beginning of this sentence is a bit unclear. A reader will likely ask, what exactly is being tested?

```

We've picked some text from this [tutorial](/engineering-education/building-a-simple-translation-app-using-python-for-beginners/) on Section's website that we want our model to summarize. 

Using our tokenizer, which we loaded earlier, let's convert our text into its tokens representation.

```python
tokens = tokenizer_model(text, truncation=True, padding="longest", return_tensors="pt")
```
We've created a variable called `tokens` to store our token representation. We use our tokenizer `tokenizer_model`, which we created earlier, to tokenize our texts. The `truncation=True` argument allows the model to truncate our texts into a size suitable for input into our model. The `return_tensors` argument tells the model to use tensors from PyTorch.

Let's view our tokens. We achieve this by writing the following command:

```python
tokens
```
Let's now go ahead and try to summarize this text.

```python

> summary = model.generate(**tokens)

Author: Solved
